src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.cs

```diff
@@ -14,6 +18,70 @@ internal static void VerifyValidPath(string path, string argName)
             }
         }
 
+        internal static bool Copy(string sourcePath, string destinationPath, bool recursive,
+            bool skipExistingFiles, CancellationToken cancellationToken)
+        {
+            SearchOption searchOption = recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
+            IEnumerable<string> directoryEnumeration;
+            bool success = true;
+
+            try
+            {
+                EnumerationOptions enumerationOptions = EnumerationOptions.Compatible;
+
+                if (recursive)
+                {
+                    enumerationOptions = EnumerationOptions.CompatibleRecursive;
+                }
+
+                directoryEnumeration = Directory.EnumerateDirectories(sourcePath, "*", enumerationOptions);
```

Comment:
Why you only read directories first? You can read files and directories and do the copy in the same loop. I suggest you to take a look at how is made here:
https://github.com/dotnet/runtime/blob/440dfe4a7beecd7755767aa247f47af00b119383/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs#L463